### DISCUSSION TITLE
why the error "more than one row returned by a subquery used as an expression" on insert operation?

### DISCUSSION THREAD

**Discussion by ricardosikic** [supabase/supabase#11900] [Questions]
Hello,
I am doing a record insert to the `initiative_categories` table. The first time it is created it works fine and creates the record but if the same registered user redoes the initiative creation flow and then adds the relationship to `initiative_categories` it returns me:
`more than one row returned by a subquery used as an expression`.

```
const { data, error } = await supabase
  .from('initiative_categories')
  .insert(initiativeCategory);
```

**Comment by GaryAustin1** [comment]
Do you have a trigger function on the insert to that table?

**Reply by ricardosikic** [reply]
everytime an initiative is created callbacks another funtion to insert on `initiative_categories` but i write a function by my own hoping to skip the issue but it only works with security definer added.

**Reply by GaryAustin1** [reply]
If it works with security definer sounds like you have RLS with a select and it is returning more than one row and you are testing with something like = which requires just one result.

Lots of hits on google with examples of what can cause it and how to fix it...
[google it](https://www.google.com/search?q=more+than+one+row+returned+by+a+subquery+postgres&client=firefox-b-1-d&ei=PijQY96OPMiqqtsPqKuOqAU&ved=0ahUKEwjepezQ7-D8AhVIlWoFHaiVA1UQ4dUDCA8&uact=5&oq=more+than+one+row+returned+by+a+subquery+postgres&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQAzIGCAAQFhAeMgYIABAWEB4yBggAEBYQHjIGCAAQFhAeMgYIABAWEB4yBggAEBYQHjIFCAAQhgMyBQgAEIYDMgUIABCGAzIFCAAQhgM6CggAEEcQ1gQQsAM6BQgAEJECOgUIABCABEoECEEYAEoECEYYAFDxBVjpEWCLFGgBcAF4AIABfIgBkgiSAQMwLjmYAQCgAQHIAQjAAQE&sclient=gws-wiz-serp
)

**Reply by ricardosikic** [reply]
yes, i saw a few examples on google but i dont have a subquery. My RLS for insert operation is: 
```
create policy "Individuals can create initiative_categories." on initiative_categories for
    insert with check (auth.uid() = (select user_id from initiatives where user_id = auth.uid()));
```
but it change the error to: `new row violates row-level security policy for table \"initiative_categories\`
thinking to work with no RLS and make validations at front haha

**Reply by GaryAustin1** [reply]
This is probably returning more than one row (this is the subquery) `auth.uid() = (select user_id from initiatives where user_id = auth.uid())`
You could use `limit 1` at the end or change the first `=` to `in`.

**Reply by GaryAustin1** [reply]
But your query is very strange.
You might want `with check (exists (select 1 from initiatives where user_id = auth.uid()))`

But you have not provided info on the table columns or your goal.

**Reply by ricardosikic** [reply]
the table of `initiative_catogories` 

```
CREATE TABLE initiative_categories (
  id bigint generated by default as identity primary key,
  category_id integer,
  initiative_id integer references public.initiatives on delete cascade not null,
  created_at timestamp with time zone default timezone('utc'::text, now()) not null,
  updated_at timestamp with time zone default timezone('utc'::text, now()) not null
);

```

**Reply by ricardosikic** [reply]
I'm thinking of only initiatives can create or update records of that table.

**Reply by ricardosikic** [reply]
nop nothing haha even if i create a new one directly on policies:

```
(initiative_id = ( SELECT 1
   FROM initiatives
  WHERE ((initiatives.id = initiative_categories.initiative_id) AND (uid() = initiatives.user_id))))
```

**Reply by GaryAustin1** [reply]
Not really sure what you are up to, but SELECT 1 returns the number 1 for each row that matches.  Not sure how you a comparing that an id...

**Reply by ricardosikic** [reply]
yes, originally the select returns only id
